src/transformers/models/bert/modeling_bert.py

```diff
@@ -375,6 +381,111 @@ def forward(
+        attention_mask: Optional[torch.Tensor] = None,
+        head_mask: Optional[torch.FloatTensor] = None,
+        encoder_hidden_states: Optional[torch.FloatTensor] = None,
+        encoder_attention_mask: Optional[torch.FloatTensor] = None,
+        past_key_value: Optional[Tuple[Tuple[torch.FloatTensor]]] = None,
+        output_attentions: Optional[bool] = False,
+    ) -> Tuple[torch.Tensor]:
+        if self.position_embedding_type != "absolute" or output_attentions or head_mask is not None:
+            # TODO: Improve this warning with e.g. `model.config._attn_implementation = "manual"` once implemented.
+            logger.warning_once(
+                "BertSdpaSelfAttention is used but `torch.nn.functional.scaled_dot_product_attention` does not support "
+                "non-absolute `position_embedding_type` or `output_attentions=True` or `head_mask`. Falling back to "
+                "the manual attention implementation, but specifying the manual implementation will be required from "
+                "Transformers version v5.0.0 onwards. This warning can be removed using the argument "
+                '`attn_implementation="eager"` when loading the model.'
+            )
+            return super().forward(
+                hidden_states,
+                attention_mask,
+                head_mask,
+                encoder_hidden_states,
+                encoder_attention_mask,
+                past_key_value,
+                output_attentions,
+            )
+
+        bsz, tgt_len, _ = hidden_states.size()
+
+        mixed_query_layer = self.query(hidden_states)
+
+        # If this is instantiated as a cross-attention module, the keys
+        # and values come from an encoder; the attention mask needs to be
+        # such that the encoder's padding tokens are not attended to.
+        is_cross_attention = encoder_hidden_states is not None
+
+        if is_cross_attention and past_key_value is not None:
+            # reuse k,v, cross_attentions
+            key_layer = past_key_value[0]
+            value_layer = past_key_value[1]
+            attention_mask = encoder_attention_mask
+        elif is_cross_attention:
+            key_layer = self.transpose_for_scores(self.key(encoder_hidden_states))
+            value_layer = self.transpose_for_scores(self.value(encoder_hidden_states))
+            attention_mask = encoder_attention_mask
+        elif past_key_value is not None:
+            key_layer = self.transpose_for_scores(self.key(hidden_states))
+            value_layer = self.transpose_for_scores(self.value(hidden_states))
+            key_layer = torch.cat([past_key_value[0], key_layer], dim=2)
+            value_layer = torch.cat([past_key_value[1], value_layer], dim=2)
+        else:
+            key_layer = self.transpose_for_scores(self.key(hidden_states))
+            value_layer = self.transpose_for_scores(self.value(hidden_states))
+
+        query_layer = self.transpose_for_scores(mixed_query_layer)
```

Comment:
```suggestion
        current_states = encoder_hidden_states if encoder_hidden_states is not None else hidden_states
        # check `seq_length` of `past_key_value` ==  `len(encoder_hidden_states)` to support prefix tuning
        attention_mask = encoder_attention_mask if is_cross_attention else attention_mask
        if is_cross_attention and past_key_value and past_key_value[0].shape[2] == current_states.shape[1]:
            key_states,  value_states = past_key_value
        else:
            key_states = self. transpose_for_scores(self.key(current_states))
            value_states = self. transpose_for_scores(self.value(current_states))
            if past_key_value is not None and not is_cross_attention:
                key_states = torch.cat([past_key_value[0], key_states], dim=2)
                value_states = torch.cat([past_key_value[1], value_states], dim=2)
```
This does not need to be so complexe. Taken from `UMT5`: we are adding a new class let's make sure we add it well.

Reply:
Sure, if that's something you feel is necessary. I agree that it is simpler, and looks to have the same logic (except for the added support prefix tuning). 

FWIW this method was adapted from `BertSelfAttention.forward()` (as the comment up top says) and those if-statements were copied directly from there with no changes. I personally prefer this approach since we minimize the risks of introducing new bugs from refactors. 

If we wanted to do a refactor, might be good for it to be in a follow-up change - and we could potentially take that opportunity to add support for the new KV cache class during the refactor as well.

Reply:
If you plan on doing the refactor for sure! 🤗 it's just that this is gonna be copied to a lot of places so I'd rather we make it right if you don't have time to come back to it and while we are at it!

Reply:
I've updated to use the refactored code. I didn't just accept the suggestion since there were a few minor differences in it (ex. key_layer instead of key_states)